### ISSUE TITLE
Weirdness when adding log vdevs to pool with 0.8.0-rc4

### ISSUE THREAD

**Issue by zviratko** [openzfs/zfs#8666]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| Ubuntu
Distribution Version	| 18.04.2
Linux Kernel	| 4.18.0-18-generic
Architecture	| amd64
ZFS Version	| 0.8.0-rc4
SPL Version	| 0.8.0-rc4

### Describe the problem you're observing
I wanted to experiment with adding logs to a pool to alleviate what I see as a "syncing" problem with many snapshots.

The pool looked like this and was likely created with some 0.7.0-rc version of ZoL, or maybe even 0.6.5.x

```  pool: backup
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
        still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
        the pool may no longer be accessible by software that does not support
        the features. See zpool-features(5) for details.
  scan: resilvered 21K in 0 days 00:30:30 with 0 errors on Thu Apr 18 10:41:47 2019
config:

        NAME                        STATE     READ WRITE CKSUM
        backup                      ONLINE       0     0     0
          mirror-0                  ONLINE       0     0     0
            scsi-350000c0f01352278  ONLINE       0     0     0
            scsi-350000c0f01e8496c  ONLINE       0     0     0
          mirror-1                  ONLINE       0     0     0
            wwn-0x50000c0f01f53d68  ONLINE       0     0     0
            scsi-350000c0f01ff280c  ONLINE       0     0     0
          mirror-2                  ONLINE       0     0     0
            wwn-0x50000c0f023d0ef4  ONLINE       0     0     0
            wwn-0x50000c0f023d17cc  ONLINE       0     0     0
          mirror-3                  ONLINE       0     0     0
            scsi-350000c0f023d1884  ONLINE       0     0     0
            wwn-0x50000c0f02effcbc  ONLINE       0     0     0
          mirror-4                  ONLINE       0     0     0
            scsi-350000c0f02f0086c  ONLINE       0     0     0
            scsi-350000c0f02f00b14  ONLINE       0     0     0
          mirror-5                  ONLINE       0     0     0
            wwn-0x5000c500b4915a70  ONLINE       0     0     0
            wwn-0x5000c500b36194a7  ONLINE       0     0     0
          mirror-6                  ONLINE       0     0     0
            wwn-0x5000cca255d63975  ONLINE       0     0     0
            wwn-0x5000cca255d63f4c  ONLINE       0     0     0
          mirror-7                  ONLINE       0     0     0
            wwn-0x5000cca255d63f18  ONLINE       0     0     0
            wwn-0x5000cca255d6397b  ONLINE       0     0     0
          mirror-8                  ONLINE       0     0     0
            wwn-0x5000cca255d60fc2  ONLINE       0     0     0
            wwn-0x5000cca271c2e0bb  ONLINE       0     0     0
          mirror-9                  ONLINE       0     0     0
            wwn-0x5000cca3b7cf0dac  ONLINE       0     0     0
            wwn-0x5000cca254e30cfa  ONLINE       0     0     0
        cache
          sda3                      ONLINE       0     0     0
          sdl3                      ONLINE       0     0     0

errors: No known data errors
```

I wanted to remove the cache devices and add them as logs:

```root@zfsbackup1:~# zpool remove backup sda3
root@zfsbackup1:~# zpool remove backup sdl3
root@zfsbackup1:~# zpool add backup log mirror sda3 sdl3
cannot add to 'backup': pool must be upgraded to add these vdevs
```
^ what, why?
But OK, I can upgrade the pool as I'm not likely to downgrade anymore.

```# zpool upgrade backup
This system supports ZFS pool feature flags.

Enabled the following features on 'backup':
  encryption
  project_quota
  device_removal
  obsolete_counts
  zpool_checkpoint
  spacemap_v2
  allocation_classes
  resilver_defer
  bookmark_v2
```

and sure enough, this now works:
```
# zpool add backup log mirror sda3 sdl3
```

```
# zpool status
  pool: backup
 state: ONLINE
  scan: resilvered 21K in 0 days 00:30:30 with 0 errors on Thu Apr 18 10:41:47 2019
config:

        NAME                        STATE     READ WRITE CKSUM
        backup                      ONLINE       0     0     0
          mirror-0                  ONLINE       0     0     0
            scsi-350000c0f01352278  ONLINE       0     0     0
            scsi-350000c0f01e8496c  ONLINE       0     0     0
          mirror-1                  ONLINE       0     0     0
            wwn-0x50000c0f01f53d68  ONLINE       0     0     0
            scsi-350000c0f01ff280c  ONLINE       0     0     0
          mirror-2                  ONLINE       0     0     0
            wwn-0x50000c0f023d0ef4  ONLINE       0     0     0
            wwn-0x50000c0f023d17cc  ONLINE       0     0     0
          mirror-3                  ONLINE       0     0     0
            scsi-350000c0f023d1884  ONLINE       0     0     0
            wwn-0x50000c0f02effcbc  ONLINE       0     0     0
          mirror-4                  ONLINE       0     0     0
            scsi-350000c0f02f0086c  ONLINE       0     0     0
            scsi-350000c0f02f00b14  ONLINE       0     0     0
          mirror-5                  ONLINE       0     0     0
            wwn-0x5000c500b4915a70  ONLINE       0     0     0
            wwn-0x5000c500b36194a7  ONLINE       0     0     0
          mirror-6                  ONLINE       0     0     0
            wwn-0x5000cca255d63975  ONLINE       0     0     0
            wwn-0x5000cca255d63f4c  ONLINE       0     0     0
          mirror-7                  ONLINE       0     0     0
            wwn-0x5000cca255d63f18  ONLINE       0     0     0
            wwn-0x5000cca255d6397b  ONLINE       0     0     0
          mirror-8                  ONLINE       0     0     0
            wwn-0x5000cca255d60fc2  ONLINE       0     0     0
            wwn-0x5000cca271c2e0bb  ONLINE       0     0     0
          mirror-9                  ONLINE       0     0     0
            wwn-0x5000cca3b7cf0dac  ONLINE       0     0     0
            wwn-0x5000cca254e30cfa  ONLINE       0     0     0
        logs
          mirror-10                 ONLINE       0     0     0
            sda3                    ONLINE       0     0     0
            sdl3                    ONLINE       0     0     0
```

This (unsurprisingly) didn't help, so I wanted to remove them and turn them into L2ARC again and I hit something that wasn't obvious at all (not directly related to this issue but I want to not it here):

```
# zpool remove backup sdl3
cannot remove sdl3: operation not supported on this type of pool
```

```
# zpool remove backup sdl3 sda3
cannot remove sdl3: operation not supported on this type of pool
cannot remove sda3: operation not supported on this type of pool
```

^ What "type of pool" is that?
Yes, I did figure it out eventually (you need to remove the "mirror-10" vdev), but this message makes no sense.

TL;DR Is this supposed to work without upgrading the pool?
```
root@zfsbackup1:~# zpool add backup log mirror sda3 sdl3
cannot add to 'backup': pool must be upgraded to add these vdevs
```

**Comment by behlendorf** [comment]
> TL;DR Is this supposed to work without upgrading the pool?

Absolutely, you shouldn't have needed to upgrade the pool.  Thanks for reporting this.  My best guess is you ended up incorrectly hitting a [this](https://github.com/zfsonlinux/zfs/blob/master/module/zfs/vdev.c#L681L698) new bit of code which was added to support allocation classes.

@don-brady it looks like this is caused by `make_leaf_vdev()` adding `ZPOOL_CONFIG_ALLOCATION_BIAS` to the config nvlist for log vdevs.  Do you have any suggestions for the best way to handle this.

**Comment by rlaager** [comment]
I've boldly added the 0.8.0 milestone so this doesn't get missed, since this is a regression and appears to be in new-to-0.8.0 code. It's easy enough to remove if this is going to be deferred.

**Comment by behlendorf** [comment]
@rlaager thanks for rounding these up!  I did mean to add this was to the 0.8 milestone.  @don-brady any thoughts on this?

**Comment by pfmooney** [comment]
This also hit us in illumos (SmartOS specifically) after the allocation classes work was [ported over](https://github.com/illumos/illumos-gate/commit/663207adb1669640c01c5ec6949ce78fd806efae).

**Comment by rugubara** [comment]
Also stumbled upon this. I've removed the log device from my pool to re-partition the disk used for log. I can no longer add it back. I run 
filename:       /lib/modules/4.19.57-gentoo/extra/zfs/zfs.ko
version:        0.8.1-r0-gentoo

Is there a fix/workaround?

**Comment by rugubara** [comment]
Since I intended to add a special vdev to the pool anyway, I have enabled feature@allocation_classes. I could immediately add a log vdev to the pool. This feature is read-only compatible, so I am still able to boot from the pool with grub.

**Comment by h1z1** [comment]
Hit this as part of some testing on an upgrade from 0.7.11 to 0.8.0.   Write IO absolutely tanked after upgrading the volume and re-adding a log.   More importantly the write IO was pegged for about an hour on reboot in some thread called zfs_upgrade?

**Comment by beren12** [comment]
Looks like this regression is still biting people.

**Comment by DeHackEd** [comment]
Can confirm 0.8.3-1el6 is still affected, and that enabling only `feature@allocation_classes` allows the log device to be added, but the feature remains only enabled (as opposed to active)

**Comment by dmaziuk** [comment]
+1: freshly upgraded to centos 7.8/zfs-0.8.4. The pool was actually set up just before centos 7.8 came out, with 2 mirrored spinning rust disks only. Now that I've upgraded, I tried adding SSD bits and got the same error message when adding log vdev:
```
cannot add to 'tank': pool must be upgraded to add these vdevs
```
Running `zpool upgrade` doesn't help.

Adding cache works however:
```
# zpool add tank cache /dev/disk/by-id/ata-Samsung_SSD_840_EVO_250GB_S1DDNEAF805308D-part4
```
-- no error.

(Also it doesn't happen when creating pools from scratch on 0.8.4, I just made 2 w/ no issues)

**Comment by brilthor** [comment]
seeing this in ubuntu 20.04 as well
```
zfs-0.8.3-1ubuntu12.4
zfs-kmod-0.8.3-1ubuntu12.4
```
as well

**Comment by Yneeb** [comment]
I think this should be reopened. I just ran into this issue using 2.1.2 on Debian.

**Comment by behlendorf** [comment]
Thanks for letting us know.  I'll go ahead and reopen this.

**Comment by dan-arnold** [comment]
I'm running into this with 2.2.3 on gentoo. Pool doesn't have feature@allocation_classes. If I want to maintain compatibility with an older version of OpenZFS that doesn't have feature@allocation_classes, is it still safe to turn it on? Or do I have to boot into the old version to add a log device?

**Comment by jspiros** [comment]
Just ran into this with 2.0.3-9+deb11u1 on Debian oldstable using a newly created version 28 pool with no feature flags. I was migrating filesystems from another, much older, version 28 pool and aiming to move the l2arc and slog over as well, but this kept me from moving the slog over.
